Core/clim-basic/extended-streams/recording.lisp

```diff
@@ -143,44 +143,47 @@ all of FUNCTION-ARGS as APPLY arguments."
 ;;; the similarities, it's necessary to treat the slots of the second
 ;;; record like variables, so for convenience the macro will use
 ;;; SLOT-VALUE on both records.
-
 (defmacro defrecord-predicate (record-type slots &body body)
   "Each element of SLOTS is either a symbol or (:initarg-name slot-name)."
-  (let* ((slot-names (mapcar #'(lambda (slot-spec)
-                                 (if (consp slot-spec)
-                                     (cadr slot-spec)
-                                     slot-spec))
-                             slots))
-         (supplied-vars (mapcar #'(lambda (slot)
-                                    (gensym (symbol-name
-                                             (symbol-concat slot '#:-p))))
-                                slot-names))
-         (key-args (mapcar #'(lambda (slot-spec supplied)
-                               `(,slot-spec nil ,supplied))
-                           slots supplied-vars))
-         (key-arg-alist (mapcar #'cons slot-names supplied-vars)))
+  (let ((key-args '())
+        (key-arg-alist '())
+        (class-slots '()))
+    (loop for spec in slots
+          for (keyword name) = (if (consp spec)
+                                   spec
+                                   (list (alexandria:make-keyword spec)
+                                         spec))
+          for suppliedp = (gensym (format nil "~A-P" keyword))
+          for slotp = (if (and (consp spec) (>= (length spec) 3))
+                          (third spec)
+                          t)
+          when slotp do (push name class-slots)
+          do (push `((,keyword ,name) nil ,suppliedp) key-args)
+          do (push (cons name suppliedp) key-arg-alist))
     `(progn
        (defmethod output-record-equal and ((record ,record-type)
                                            (record2 ,record-type))
-         (macrolet ((if-supplied ((var &optional (type t)) &body supplied-body)
-                      (declare (ignore var type))
-                      `(progn ,@supplied-body)))
-           (with-slots ,slot-names record2
+         (macrolet ((if-supplied ((var &optional (type t))
+                                  slot-form &optional arg-form)
```

Comment:
I think that changing the if-supplied macro lambda list doesn't do much good (i.e it is less comprehensible). The test for the text-output record could be expressed as:

```
(if-supplied (string)
  (let ((start (or (if-supplied (start) start) 0))
        (end (or (if-supplied (end) end) nil)))
    (string= (slot-value record 'string) string
             :start2 start :end end)))
```

Reply:
I'm not sure I understand this suggestion. Wouldn't that try to access `start` and `end` as slots in the method that compares two output record objects?

Reply:
in your code (find var ',class-slots :test #'eq) will yield nil, so the body won't be expanded,

(or (if-supplied (start) start) 0) will expand to
(or nil 0)